src/java.base/share/classes/sun/security/provider/Dilithium.java

```diff
@@ -0,0 +1,1432 @@
+        dilithiumH.update(randomBytes);
+        dilithiumH.update((byte)dilithium_k);
+        dilithiumH.update((byte)dilithium_l);
+        byte[] rho = dilithiumH.squeeze(dilithiumASeedLength);
+        byte[] rhoPrime = dilithiumH.squeeze(dilithiumS1S2SeedLength);
+        byte[] k = dilithiumH.squeeze(dilithiumKLength);
+        dilithiumH.reset();
+
+        //Sample A
+        int[][][] keygenA = generateA(rho); //A is in NTT domain
+
+        //Sample S1 and S2
+        int[][] s1 = new int[dilithium_l][dilithium_n];
+        int[][] s2 = new int[dilithium_k][dilithium_n];
+        sampleS1S2(s1, s2, dilithiumH, rhoPrime); //dilithiumH is reset before being used in sampleS1S2
+
+        //Compute t and tr
+        mlDsaVectorNtt(s1); //s1 now in NTT domain
+        int[][] As1 = matrixVectorPointwiseMultiply(keygenA, s1);
+        mlDsaVectorInverseNtt(s1); //take s1 out of NTT domain
+
+        mlDsaVectorInverseNtt(As1);
+        int[][] t = vectorAddPos(As1, s2);
+        int[][] t0 = new int[dilithium_k][dilithium_n];
+        int[][] t1 = new int[dilithium_k][dilithium_n];
+        power2Round(t, t0, t1);
+
+        //Encode PK and SK
+        DilithiumPublicKey pk = new DilithiumPublicKey(rho, t1);
+        byte[] publicKeyBytes = pkEncode(pk);
+        dilithiumCRH.update(publicKeyBytes);
+        byte[] tr = dilithiumCRH.digest();
+        DilithiumPrivateKey sk = new DilithiumPrivateKey(rho, k, tr, s1, s2, t0);
+
+        return new DilithiumKeyPair(sk, pk);
+    }
+
+    public DilithiumSignature sign(byte[] message, byte[] rnd, DilithiumPrivateKey sk) {
+        //Initialize hash functions
+        var dilithiumH = new SHAKE256(0);
+
+        int[][] s1 = Arrays.stream(sk.s1()).map(int[]::clone).toArray(int[][]::new);
+        int[][] s2 = Arrays.stream(sk.s2()).map(int[]::clone).toArray(int[][]::new);
+        int[][] t0 = Arrays.stream(sk.t0()).map(int[]::clone).toArray(int[][]::new);
+
+        //Do some NTTs
+        mlDsaVectorNtt(s1);
+        mlDsaVectorNtt(s2);
+        mlDsaVectorNtt(t0);
+        int[][][] aHat = generateA(sk.rho());
+
+        //Compute mu
+        dilithiumH.update(sk.tr());
+        dilithiumH.update(message);
+        byte[] mu = dilithiumH.squeeze(dilithiumMuLength);
+        dilithiumH.reset();
+
+        //Compute rho'
+        dilithiumH.update(sk.k());
+        dilithiumH.update(rnd);
+        dilithiumH.update(mu);
+        byte[] rhoPrime = dilithiumH.squeeze(dilithiumMaskSeedLength);
+        dilithiumH.reset();
+
+        //Initialize vectors used in loop
+        int[][] z = null;
+        boolean[][] h = null;
+        byte[] commitmentHash = new byte[lambda/4];
+
+        int kappa = 0;
+        while (true) {
+            int[][] y = expandMask(rhoPrime, kappa);
+
+            //Compute w and w1
+            mlDsaVectorNtt(y); //y is now in NTT domain
+            int[][] w = matrixVectorPointwiseMultiply(aHat, y);
+            mlDsaVectorInverseNtt(w); //w is not in normal domain
```

Comment:
If you modify methods like `mlDsaVectorInverseNtt` and `mlDsaVectorInverseNtt` to return the input argument, the 3 lines above can be written into
```
var w = mlDsaVectorInverseNtt(matrixVectorPointwiseMultiply(aHat, mlDsaVectorNtt(y)));
```
which is very similar to the original algorithm.